File: rust/tests/playback.rs
//! End-to-end proof of the FFI chain: Rust fixture → native VCR
//! (aether_vcr_embed_*) → embedded Aether HTTP server. Replays a Servirtium
//! markdown tape and asserts the SUT-visible response and diagnostics.

mod common;

use common::{raw_get, tape_path};
use servirtium::{Field, Outcome, Vcr};

#[test]
fn replays_a_recorded_get_on_a_dynamic_port() {
    let vcr = Vcr::playback(tape_path("single_get.md"))
        .label("replays a recorded GET")
        .port(0)
        .start()
        .expect("playback should start");

    assert!(vcr.port() > 0, "expected an OS-assigned port");
    assert_eq!(1, vcr.tape_length());

    let body = ureq::get(&format!("{}/ok", vcr.base_url()))
        .call()
        .expect("request should succeed")
        .into_string()
        .unwrap();

    assert_eq!("ok-body", body);
    assert_eq!(Outcome::Ok, vcr.last_kind());
    assert_eq!("", vcr.last_error());
}

#[test]
fn flags_a_path_mismatch_via_diagnostics() {
    let vcr = Vcr::playback(tape_path("single_get.md")).port(0).start().unwrap();

    // 599-style mismatch; ureq surfaces it as an error status — ignore the
    // body, assert on the diagnostics.
    let _ = ureq::get(&format!("{}/nope", vcr.base_url())).call();

    assert_ne!(Outcome::Ok, vcr.last_kind());
    assert!(!vcr.last_error().is_empty(), "expected a mismatch diagnostic");
}

#[test]
fn unredaction_lets_a_scrubbed_tape_match_the_real_request() {
    let vcr = Vcr::playback(tape_path("secure_get.md"))
        .strict_headers()
        .unredact(Field::RequestHeaders, "Bearer REDACTED", "Bearer real-token")
        .port(0)
        .start()
        .unwrap();

    // A precise request (only Host + Authorization) so strict matching has
    // exactly the recorded header set to compare against.
    let body = raw_get(&vcr.base_url(), "/secure", &[("Authorization", "Bearer real-token")]);

    assert_eq!("secret-ok", body);
    assert_eq!(Outcome::Ok, vcr.last_kind());
}

#[test]
fn strict_matching_flags_a_missing_request_header() {
    let vcr = Vcr::playback(tape_path("secure_get.md"))
        .strict_headers()
        .unredact(Field::RequestHeaders, "Bearer REDACTED", "Bearer real-token")
        .port(0)
        .start()
        .unwrap();

    // No Authorization header at all → mismatch.
    let _ = raw_get(&vcr.base_url(), "/secure", &[]);

    assert_ne!(Outcome::Ok, vcr.last_kind());
    assert!(!vcr.last_error().is_empty());
}

#[test]
fn static_content_is_served_from_disk_not_the_tape() {
    let dir = std::env::temp_dir().join(format!(
        "vcr_static_{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos()
    ));
    std::fs::create_dir_all(&dir).unwrap();
    std::fs::write(dir.join("asset.txt"), "static-asset").unwrap();

    let vcr = Vcr::playback(tape_path("single_get.md"))
        .static_content("/files", dir.to_string_lossy().to_string())
        .port(0)
        .start()
        .unwrap();

    let base = vcr.base_url();
    let from_disk = ureq::get(&format!("{base}/files/asset.txt"))
        .call()
        .unwrap()
        .into_string()
        .unwrap();
    let from_tape = ureq::get(&format!("{base}/ok"))
        .call()
        .unwrap()
        .into_string()
        .unwrap();

    assert_eq!("static-asset", from_disk);
    assert_eq!("ok-body", from_tape);

    std::fs::remove_dir_all(&dir).ok();
}

File: lua/tests/playback.lua
-- playback.lua — the Lua binding's playback test suite, mirroring the Go
-- binding's playback_test.go. Drives real HTTP with curl against playback
-- servers (one server per port). Run from lua/ with:
--   LUA_CPATH="./?.so;;" SERVIRTIUM_VCR_LIB=<abs>/libservirtium_vcr.so lua5.4 tests/playback.lua

local here = arg[0]:match("^(.*)[/\\]") or "."
package.path  = here .. "/../?.lua;" .. here .. "/?.lua;" .. package.path
package.cpath = here .. "/../?.so;" .. package.cpath

local servirtium = require("servirtium")
local h = require("helpers")

local function tape(name) return h.tapes_dir .. "/" .. name end

-- replays a recorded GET ------------------------------------------------------
do
    h.case("replays a recorded GET")
    local srv = assert(servirtium.playback(tape("single_get.md"))
        :label("replays a recorded GET"):port(0):start())

    h.truthy(srv:port() > 0, "OS-assigned port")
    h.eq(srv:tape_length(), 1, "tape length")

    local body = h.curl_get(srv:base_url(), "/ok")
    h.eq(body, "ok-body", "body")
    h.eq(srv:last_kind(), servirtium.OK, "last_kind")
    h.eq(srv:last_error(), "", "last_error")
    srv:close()
end

-- flags a path mismatch via diagnostics --------------------------------------
do
    h.case("flags a path mismatch")
    local srv = assert(servirtium.playback(tape("single_get.md")):port(0):start())
    h.curl_get(srv:base_url(), "/nope")
    h.truthy(srv:last_kind() ~= servirtium.OK, "mismatch outcome (not Ok)")
    h.truthy(srv:last_error() ~= "", "mismatch diagnostic non-empty")
    srv:close()
end

-- unredaction lets a scrubbed tape match -------------------------------------
do
    h.case("unredaction lets a scrubbed tape match")
    local srv = assert(servirtium.playback(tape("secure_get.md"))
        :strict_headers()
        :unredact(servirtium.FIELD_REQUEST_HEADERS, "Bearer REDACTED", "Bearer real-token")
        :port(0):start())

    -- curl sends a default User-Agent; the strict tape only records
    -- Authorization, so suppress it (-H "User-Agent:") to match the block.
    local p = io.popen("curl -s -H 'Authorization: Bearer real-token' " ..
        "-H 'User-Agent:' -H 'Accept:' " .. srv:base_url() .. "/secure")
    local body = p:read("*a"); p:close()

    h.eq(body, "secret-ok", "body")
    h.eq(srv:last_kind(), servirtium.OK, "last_kind (" .. srv:last_error() .. ")")
    srv:close()
end

-- strict matching flags a missing request header -----------------------------
do
    h.case("strict matching flags a missing request header")
    local srv = assert(servirtium.playback(tape("secure_get.md"))
        :strict_headers()
        :unredact(servirtium.FIELD_REQUEST_HEADERS, "Bearer REDACTED", "Bearer real-token")
        :port(0):start())

    -- No Authorization header at all -> mismatch.
    h.curl_get(srv:base_url(), "/secure")
    h.truthy(srv:last_kind() ~= servirtium.OK, "mismatch outcome (not Ok)")
    h.truthy(srv:last_error() ~= "", "mismatch diagnostic non-empty")
    srv:close()
end

-- static content served from disk --------------------------------------------
do
    h.case("static content served from disk")
    local dir = h.temp_tape("assets"):gsub("%.md$", "")
    os.execute("mkdir -p " .. dir)
    local f = assert(io.open(dir .. "/asset.txt", "w"))
    f:write("static-asset"); f:close()

    local srv = assert(servirtium.playback(tape("single_get.md"))
        :static_content("/files", dir):port(0):start())

    h.eq(h.curl_get(srv:base_url(), "/files/asset.txt"), "static-asset", "from disk")
    h.eq(h.curl_get(srv:base_url(), "/ok"), "ok-body", "from tape (unaffected)")
    srv:close()
    os.execute("rm -rf " .. dir)
end

-- untaped path 404s without consuming the cursor -----------------------------
do
    h.case("untaped 404s without consuming the cursor")
    local srv = assert(servirtium.playback(tape("single_get.md"))
        :untaped("/favicon.ico"):port(0):start())

    h.eq(h.curl_status(srv:base_url(), "/favicon.ico"), 404, "untaped 404")
    -- The normal recorded interaction still replays (cursor wasn't consumed).
    h.eq(h.curl_get(srv:base_url(), "/ok"), "ok-body", "tape still replays")
    h.eq(srv:last_kind(), servirtium.OK, "last_kind after untaped")
    srv:close()
end

-- two playback servers at once (one server per port) -------------------------
do
    h.case("two playback servers at once")
    local a = assert(servirtium.playback(tape("single_get.md")):port(0):start())
    local b = assert(servirtium.playback(tape("secure_get.md")):port(0):start())

    h.truthy(a:port() ~= b:port(), "distinct ports")
    h.eq(h.curl_get(a:base_url(), "/ok"), "ok-body", "server A from its tape")

    local p = io.popen("curl -s -H 'Authorization: Bearer REDACTED' " ..
        "-H 'User-Agent:' -H 'Accept:' " .. b:base_url() .. "/secure")
    h.eq(p:read("*a"), "secret-ok", "server B from its tape"); p:close()

    a:close(); b:close()
end

h.done("playback")
